setup.py

```diff
@@ -1,48 +1,88 @@
 from __future__ import absolute_import, division, print_function
 
-from setuptools import setup, find_packages, Extension
+import glob
 import os
+import sys
+from setuptools import setup, find_packages, Extension
+from setuptools.command.build_ext import build_ext as _build_ext
 
-import numpy as np
+# Python 2 has a different name for builtins.
+try:
+    import builtins
+except ImportError:
+    import __builtin__ as builtins
 
-from Cython.Build import cythonize
+try:
+    from Cython.Build import cythonize
+except ImportError:
+    print('Warning: Cython is not available - '
+          'will be unable to build stratify extensions')
+    cythonize = None
 
+PACKAGE_NAME = 'stratify'
+PACKAGE_DIR = os.path.abspath(os.path.dirname(__file__))
+CMDS_NOCYTHONIZE = ['clean', 'sdist']
 
-NAME = 'stratify'
-DIR = os.path.abspath(os.path.dirname(__file__))
 
-extension_kwargs = {'include_dirs': [np.get_include()]}
-cython_coverage_enabled = os.environ.get('CYTHON_COVERAGE', None)
-if cython_coverage_enabled:
-    extension_kwargs.update({'define_macros': [('CYTHON_TRACE_NOGIL', '1')]})
+class NumpyBuildExt(_build_ext):
+    # Delay numpy import so that setup.py can be run without numpy already
+    # being installed.
+    def finalize_options(self):
+        _build_ext.finalize_options(self)
+        builtins.__NUMPY_SETUP__ = False
+        import numpy
+        self.include_dirs.append(numpy.get_include())
 
-extensions = [Extension('{}._vinterp'.format(NAME),
-                        [os.path.join(NAME, '_vinterp.pyx')],
-                        **extension_kwargs),
-              Extension('{}._conservative'.format(NAME),
-                        [os.path.join(NAME, '_conservative.pyx')],
-                        **extension_kwargs)]
 
 def extract_version():
     version = None
-    fname = os.path.join(DIR, NAME, '__init__.py')
+    fname = os.path.join(PACKAGE_DIR, PACKAGE_NAME, '__init__.py')
     with open(fname) as fd:
         for line in fd:
-            if (line.startswith('__version__')):
+            if line.startswith('__version__'):
                 _, version = line.split('=')
                 version = version.strip()[1:-1]  # Remove quotations
                 break
     return version
 
 
+# Python 2 is not supported by numpy as of version 1.17
+# but pip will attempt to install/use newer Python 3-only numpy versions.
+if sys.version_info.major < 3:
+    numpy_req = 'numpy<1.17'
+else:
+    numpy_req = 'numpy<2.0'
+requirements = ['setuptools>=18.0', numpy_req, 'Cython']
+
+extension_kwargs = {}
+cython_directives = {'binding': True}
```

Comment:
@tjtg Is it really necessary to enable the `binding` Cython directive? From the [documentation](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiler-directives), I don't see the direct benefit. Just wondered if you could explain the advantage of this, thanks :wink:

Reply:
`binding` was previously enabled so this is intended to leave the configuration as-is - see https://github.com/SciTools-incubator/python-stratify/blob/0eba3a8454707afd8912204d47dc3945b7a0b7fb/setup.py#L45 for the old setting location.

Reply:
@tjtg Yup, you're right. You inherited this... so we can deal with that, but not in this PR :smile: